### PR TITLE
Allow referencing existing Defender resources

### PR DESCRIPTION
**Pull request by shahnami**
https://linear.app/openzeppelin-development/issue/PLAT-2817/allow-defender-as-code-to-reference-existing-resources

### CODE REVIEW

**Review by MCarlomagno — APPROVED** [review]
LGTM, my only concern is about security, is it possible to reference resources from another tenant? Let's say I see the actionId from another tenant in the forum and I attack that tenant triggering that action using a monitor from my own tenant, would that be possible?

File: src/cmd/deploy.ts [hunk 1]
```diff
@@ -142,8 +157,12 @@ export default class DefenderDeploy {
     const monitorDifference = _.differenceWith(
       monitorItems,
       Object.entries(monitors),
-      (a: DefenderMonitor, b: [string, Monitor]) =>
-        a.stackResourceId === getResourceID(getStackName(this.serverless), b[0]),
+      (a: DefenderMonitor, b: [string, Monitor | DefenderID]) => {
```
Comment (CoveMB) [inline src/cmd/deploy.ts:160]: nits: idealy they could be full names for readability (here and everywhere where variable are only 1 letters)

File: src/types/types/resources.schema.d.ts [hunk 2]
```diff
@@ -100,27 +106,54 @@ export type OpsgeniePriorityLevel = 'P1' | 'P2' | 'P3' | 'P4' | 'P5';
 export type PagerDutyEventType = 'change' | 'alert';
 export type PagerDutyEventAction = 'trigger' | 'acknowledge' | 'resolve';
 export type PagerDutySeverity = 'critical' | 'error' | 'warning' | 'info';
-export type CategoryNotificationIds = Notification1[];
+export type DefenderID2 = string;
+export type CategoryOrDefenderID = Category | DefenderID3;
+export type NotificationOrDefenderID1 = Notification | DefenderID2;
+export type CategoryNotificationIds = NotificationOrDefenderID1[];
+export type DefenderID3 = string;
+export type RelayerOrDefenderID1 = Relayer | DefenderID;
```
Comment (CoveMB) [inline src/types/types/resources.schema.d.ts:114]: should the "1" at the end be there?

Comment (shahnami) [reply src/types/types/resources.schema.d.ts:114]: Good catch. These are auto-generated by json-schema for documentation

**Comment by shahnami** [comment]
> LGTM, my only concern is about security, is it possible to reference resources from another tenant? Let's say I see the actionId from another tenant in the forum and I attack that tenant triggering that action using a monitor from my own tenant, would that be possible?

These are secured by tenant-scoped API keys (since it uses defender-sdk). When a reference is made, we try to pull that data in during, for example, the deploy, but if the user doesn't have access to that resource, the authoriser will throw.

File: src/cmd/deploy.ts [hunk 3]
```diff
@@ -736,12 +776,16 @@ export default class DefenderDeploy {
       const actions = await getActionClient(this.teamKey!).list();
       const notifications = await client.listNotificationChannels();
       const categories = await client.listNotificationCategories();
+
+      // TODO: Add a new endpoint in defender-sdk which includes contract ABI and NatSpec
+      const contracts = await getProposalClient(this.teamKey!).listContracts();
```
[the remark this answers is not in the document]
Comment (shahnami) [reply src/cmd/deploy.ts:781]: https://linear.app/openzeppelin-development/issue/PLAT-3170/create-a-list-contract-endpoint-that-includes-abi

**Pull request merged**